src/runtime-rs/crates/hypervisor/src/kernel_param.rs

```diff
@@ -0,0 +1,210 @@
+        ];
+
+        if debug {
+            params.push(Param {
+                key: "agent.log_vport".to_string(),
+                value: VSOCK_LOGS_PORT.to_string(),
+            });
+        }
+
+        Self { params }
+    }
+
+    pub(crate) fn new_rootfs_kernel_params(rootfs_driver: &str) -> Self {
+        let params = match rootfs_driver {
+            VM_ROOTFS_DRIVER_BLK => {
+                vec![
+                    Param {
+                        key: "root".to_string(),
+                        value: "/dev/vda1".to_string(),
+                    },
+                    Param {
+                        key: "rootflags".to_string(),
+                        value: "data=ordered,errors=remount-ro ro".to_string(),
+                    },
+                    Param {
+                        key: "rootfstype".to_string(),
+                        value: "ext4".to_string(),
+                    },
+                ]
+            }
+            VM_ROOTFS_DRIVER_PMEM => {
+                vec![
+                    Param {
+                        key: "root".to_string(),
+                        value: "/dev/pmem0p1".to_string(),
+                    },
+                    Param {
+                        key: "rootflags".to_string(),
+                        value: "data=ordered,errors=remount-ro,dax ro".to_string(),
+                    },
+                    Param {
+                        key: "rootfstype".to_string(),
+                        value: "ext4".to_string(),
+                    },
+                ]
+            }
+            _ => vec![],
+        };
+        Self { params }
+    }
+
+    pub(crate) fn append(&mut self, params: &mut KernelParams) {
+        self.params.append(&mut params.params);
+    }
+
+    pub(crate) fn from_string(params_string: &str) -> Self {
+        let mut params = vec![];
+
+        let parameters_vec: Vec<&str> = params_string.split(KERNEL_PARAM_DELIMITER).collect();
+
+        for param in parameters_vec.iter() {
+            if param.is_empty() {
+                continue;
+            }
+
+            let ps: Vec<&str> = param.splitn::<_>(2, KERNEL_KV_DELIMITER).collect();
+
+            if ps.len() == 2 {
+                params.push(Param {
+                    key: String::from(ps[0]),
+                    value: String::from(ps[1]),
+                });
+            } else {
+                params.push(Param {
+                    key: String::from(ps[0]),
+                    value: String::from(""),
+                });
+            }
+        }
+
+        Self { params }
+    }
+}
+
+impl fmt::Display for KernelParams {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut parameters: Vec<String> = Vec::new();
+
+        for param in &self.params {
+            if param.key.is_empty() && param.value.is_empty() {
+                continue;
+            } else if param.key.is_empty() {
+                parameters.push(param.value.to_string());
```

Comment:
fwics the only way for this to be possible is if a param was specified as `=value`, which seems unlikely tbh ;) Shouldn't that be considered an error?

Reply:
Thanks for your review, return an error for this case.